Pathkit FFI start.

dnfield opened the pull request:
Very WIP

@c-h-i-a-m-a-k-a-2 @jonahwilliams fyi

skia-gold commented:
Gold has detected about 1 new digest(s) on patchset 4.
View them at https://flutter-engine-gold.skia.org/cl/github/34267

yjbanov commented on tools/pathkit/dart/lib/pathkit.dart:
```diff
@@ -0,0 +1,381 @@
+  } else if (Platform.isIOS || Platform.isMacOS) {
+    return ffi.DynamicLibrary.open('libpathkit.dylib');
+  } else if (Platform.isAndroid || Platform.isLinux) {
+    return ffi.DynamicLibrary.open('libpathkit.so');
```
Can we use wasm to avoid having to ship platform-specific libraries? I'm not sure what the wasm support level is like in the Dart VM.

jonahwilliams replied:
I made Dan promise he wouldn't

jonahwilliams replied:
(Not really)

dnfield replied:
In theory yes, but we'd need to ship a wasm runtime to use it and it probably won't perform quite as well.

dnfield replied:
(Skia already has a WASM version of this that they publish called "PathKit" on NPM)

dnfield commented:
- Changed applyOp to return a new path
- Renamed this to `path_ops`, because it's not really the same as the existing `pathkit` library Skia distributes.
- Added some Dart based tests.

zanderso commented on tools/path_ops/dart/lib/path_ops.dart:
```diff
@@ -0,0 +1,390 @@
+
+// TODO(dnfield): Figure out where to put this.
+// https://github.com/flutter/flutter/issues/99563
+final ffi.DynamicLibrary _dylib = () {
```
My understanding is that it is more idiomatic to import (and not re-export) the stuff from here down from a library under `lib/src`.

dnfield replied:
This is all private to the library - but I can move it.

dnfield replied:
The problem with importing it is that people could then do source imports, which I'd like to avoid. I'm not sure which is better to handle here.

This is a basic namespacing/visibility problem in Dart though and maybe I just need to accept it?  I know in the framework we don't trust people to not do source imports.

dnfield commented on tools/path_ops/BUILD.gn:
```diff
@@ -0,0 +1,35 @@
+import("//flutter/shell/version/version.gni")
```
Done

jonahwilliams reviewed (commented):
How do you intended to use this Dart library? (I cannot think of any reasonable way it could be imported besides path imports - unless there is a separate step that updates to pub)

dnfield commented:
@jonahwilliams - it's basically the same as the tessellator one.

In that case I think we more or less copy-pasted the Dart portion directly into vector_graphics_compiler.

Ideally, we'd have a way to publish FFI based libraries to pub that do not require developers to build from source (since the build system here is not CMake and building this library would take a non-trivial amount of time for "normal" developer machines without GOMA).

dnfield commented:
I'm going to land this and we can figure out the right way to consume this as we go. I think worst case we can copy/paste the Dart library, or distribute it via the Flutter SDK if we don't have a good way to distribute it via pub.
